src/Features/Core/Portable/Workspace/CompileTimeSolutionProvider.cs

```diff
@@ -116,20 +115,27 @@ public Solution GetCompileTimeSolution(Solution designTimeSolution)
                                 documentIdsToRemove.Add(documentState.Id);
                             }
                         }
+
+                        compileTimeSolution = compileTimeSolution
+                            .RemoveAnalyzerConfigDocuments(configIdsToRemove.ToImmutable())
+                            .RemoveDocuments(documentIdsToRemove.ToImmutable());
+
+                        if (staleSolution is not null)
+                        {
+                            var existingStaleProject = staleSolution.GetProject(projectState.Id);
+                            if (existingStaleProject is not null)
+                                compileTimeSolution = compileTimeSolution.WithCachedSourceGeneratorState(projectState.Id, existingStaleProject);
+                        }
                     }
                 }
 
-                var compileTimeSolution = designTimeSolution
-                    .RemoveAnalyzerConfigDocuments(configIdsToRemove.ToImmutable())
-                    .RemoveDocuments(documentIdsToRemove.ToImmutable());
```

Comment:
We were previously doing the removes as a single operation. I didn't see a huge reason to really do that, since the first thing these methods do is re-group the IDs by project and then apply the changes to each project independently.

Reply:
I suppose the only real difference is us re-acquiring the array builder in the loop now. That doesn't have an impact on allocations?

Reply:
I wouldn't imagine many allocations, no; considering this is going to be recreating compilations/symbols like crazy this is the cheap part of this!